packages/web3-core/src/web3_subscriptions.ts

```diff
@@ -71,18 +72,33 @@ export abstract class Web3Subscription<
 		});
 
 		const messageListener = (
-			err: Error | undefined,
-			data?: JsonRpcSubscriptionResult | JsonRpcNotification<Log>,
+			data?:
+				| JsonRpcSubscriptionResult
+				| JsonRpcSubscriptionResultOld<Log>
+				| JsonRpcNotification<Log>,
 		) => {
-			if (data && jsonRpc.isResponseWithNotification(data)) {
-				this._processSubscriptionResult(data?.params.result);
+			// for EIP-1193 provider
+			if (data?.data) {
+				this._processSubscriptionResult(data?.data?.result ?? data?.data);
+				return;
 			}
-			if (err) {
-				this._processSubscriptionError(err);
+
+			if (
+				data &&
+				jsonRpc.isResponseWithNotification(
+					data as unknown as JsonRpcSubscriptionResult | JsonRpcNotification<Log>,
+				)
+			) {
+				this._processSubscriptionResult(data?.params.result);
 			}
 		};
 
-		(this._requestManager.provider as Web3BaseProvider).on<Log>('message', messageListener);
+		// @ts-expect-error subscribe using old providers
+		if ((typeof this._requestManager.provider as Web3BaseProvider).request === 'function') {
+			(this._requestManager.provider as Web3BaseProvider).on<Log>('data', messageListener);
+		} else {
+			(this._requestManager.provider as Web3BaseProvider).on<Log>('message', messageListener);
+		}
```

Comment:
This is backwards, [it should be](https://github.com/web3/web3.js/blob/13a2edc852aa2ae35cf827f8b08958c26356db4c/packages/web3-core-requestmanager/src/index.js#L100-L101)

```suggestion
		if ((typeof this._requestManager.provider as Web3BaseProvider).request === 'function') {
			(this._requestManager.provider as Web3BaseProvider).on<Log>('message', messageListener);
		} else {
			(this._requestManager.provider as Web3BaseProvider).on<Log>('data', messageListener);
		}
```

Reply:
omg! yes. thank you @spacesailor24 !